src/lib/Templates.ml

```diff
@@ -120,74 +146,298 @@ let rec instantiate_instructions (param_map : var_map) : parsed_seq_code_bloc ->
+              else failwith "[expand_parsed_discrete_factor]: Index is greater or equal to length of syntatic array."
+    end
+    | _ -> factor
+
+let expand_name_or_access_list = List.map expand_name_or_access
+
+let expand_sync : unexpanded_sync -> sync = function
+  | UnexpandedSync action -> Sync (expand_name_or_access action)
+  | UnexpandedNoSync -> NoSync
+
+let expand_indexed_update (synt_vars : synt_vars_data) : parsed_scalar_or_index_update_type -> parsed_scalar_or_index_update_type =
+  fun e ->
+    match e with
+      | Parsed_indexed_update (Parsed_scalar_update (name, _), index) -> begin
+          match List.find_map (fun (name', _, len) -> if name = name' then Some len else None) synt_vars with
+            | None -> e
+            | Some len ->
+                let index_c = NumConst.to_bounded_int (get_const_disc_arith_expr index) in
+                if index_c < len then
+                  let new_name = gen_access_id name index_c in
+                  Parsed_scalar_update (new_name, 0)
+                else
+                  failwith "[expand_indexed_update]: Index is greater or equal to length of syntatic array"
+      end
+      | _ -> e
+
+let rec expand_instruction (synt_vars : synt_vars_data) : parsed_instruction -> parsed_instruction = function
+  | Parsed_local_decl (name, tp, init) ->
+      let init' = expand_parsed_boolean_expression synt_vars init in
+      Parsed_local_decl (name, tp, init')
+  | Parsed_assignment (upd, rhs) ->
+      let rhs' = expand_parsed_boolean_expression synt_vars rhs in
+      let upd' = expand_indexed_update synt_vars upd in
+      Parsed_assignment (upd', rhs')
+  | Parsed_instruction expr -> Parsed_instruction (expand_parsed_boolean_expression synt_vars expr)
+  | Parsed_for_loop (var_ref, left_expr, right_expr, dir, bloc) ->
+      let left_expr' = expand_parsed_discrete_arithmetic_expression synt_vars left_expr in
+      let right_expr' = expand_parsed_discrete_arithmetic_expression synt_vars right_expr in
+      let bloc' = expand_code_bloc synt_vars bloc in
+      Parsed_for_loop (var_ref, left_expr', right_expr', dir, bloc')
+  | Parsed_while_loop (cond, bloc) ->
+      let cond' = expand_parsed_boolean_expression synt_vars cond in
+      let bloc' = expand_code_bloc synt_vars bloc in
+      Parsed_while_loop (cond', bloc')
+  | Parsed_if (cond, then_branch, else_branch_opt) ->
+      let cond' = expand_parsed_boolean_expression synt_vars cond in
+      let then_branch' = expand_code_bloc synt_vars then_branch in
+      let else_branch_opt' = Option.map (fun expr -> expand_code_bloc synt_vars expr) else_branch_opt in
+      Parsed_if (cond', then_branch', else_branch_opt')
+
+and expand_code_bloc (synt_vars : synt_vars_data) : parsed_seq_code_bloc -> parsed_seq_code_bloc =
+  List.map (expand_instruction synt_vars)
+
+let expand_loc (synt_vars : synt_vars_data) (loc : unexpanded_parsed_location) : parsed_location =
+  let expand_transition (guard, bloc, sync, loc_name) =
```

Comment:
what is the benefit of defining functions `expand_transition` and `expand_flow` inside `expand_loc` ?

Reply:
ohh, I understood :) so it's OK :)